### DISCUSSION TITLE
Maybe error in entropy and VCI

### DISCUSSION THREAD

**Discussion by XingyanCao** [r-lidar/lidR#743] [General]
Hi Jean-Romain,

Thanks for your contribution in developing such good and useful package for computing data metrics from the ALS data.

While I have a question on the script computing entropy and vertical complexity index (VCI). I noticed that the function computing VCI returned the same value of function computing entropy did when you set zmax as the maximum canopy height and set by as 1.

I checked the script in the github and found that the function entropy divided the value by the logrithm value of the maximum canopy height, while this should be in the equation of VCI, not entropy. And in the function VCI, it just returns the value of entropy value without extra computation.

Best regards,

**Comment by Jean-Romain** [comment]
> the function entropy divided the value by the logrithm value of the maximum canopy height

I do not see that. Please cite some code and explain what you think is incorrect. I hardly understand what may be incorrect based on your explanation. Thanks

**Reply by XingyanCao** [reply]
In the file lidR/R/metrics_stdmetrics.R, line 623, you defined the function entropy(). At line 649, you return the output as 
```r
 S <- -sum(p*log(p)) / -sum(pref*log(pref)). 
```
But the entropy value should be 

```r
S <- -sum(p*log(p))
```

based on its definition. 

And at line 668, you defined function VCI and return it as `entropy(z, by, zmax)`, while based on the definition in the paper(doi.org/10.14358/PERS.77.3.261, page 3, table 1), it should be `entropy/ln(HB)`, where HB is the number of height bins.

**Reply by Jean-Romain** [reply]
The documentation states that it is a normalized entropy. 

According to the source code it computes the entropy (` -sum(p*log(p)) `) divided by the maximum possible entropy (` -sum(pref*log(pref))`).  Indeed the entropy depends on the height. Let assume that you have two bins in a two meters height region. The maximum entropy is `2*1/2*log(1/2) = log(1/2) = 0.69` . Now in a 10 meters regions the maximum entropy is `-log(1/10) = 2.3`. 

Consequently, you can see that the entropy is a metric that measures more the elevation than the dispersion. This is why it is normalized in [0,1]. This way it describes the dispersion relative to the maximum dispersion.

**Reply by Jean-Romain** [reply]
For VCI you may be right. I'm surprised, I need to check again later.

**Comment by XingyanCao** [comment]
Yes, the problem is that the script in the function entropy returns value of VCI rather than entropy. Based on the definition, VCI is entropy divided by the maximum possible entropy.

**Reply by Jean-Romain** [reply]
Yes, VCI is divided by the maximum entropy. But the maximum of which height? We have seen than the max entropy with 2 bins is 0.69. With 10 bins it is 2.3. In entropy() we divide by the maximum entropy at the maximum local height. In VCI we divide by a global height fixed by the user.

Entropy is the normalized entropy, normalized locally. VCI is normalized globally.

**Reply by XingyanCao** [reply]
Thanks for you explaination, I know it.

**Comment by XingyanCao** [comment]
Another thing may be that, if the given value for VCI is larger than the maximum canopy height, it still returns value, but that value is incorrect, I think it may be better to add an if loop to avoid such situation.
